### PR TITLE
Create all the database tables using django models   p1  cu 86enpcmpp

### PR DESCRIPTION
**Pull request by sanjueranga**

### CODE REVIEW

File: users/models.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class Experience(models.Model):
+    label = models.TextField(max_length=500)
+    def __str__(self):
+        return self.label
+
+
+class UserProfile(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    company_name = models.CharField(max_length=200, null=True, blank=True)
+    profile_picture = models.URLField(max_length=100, blank=True, null=True)
```
Comment (TWijesooriya) [inline users/models.py]: Generally when handling image uploads, we upload images to a specific location on the server chosen by us. So as we already know the exact location where all the images live, it's enough for us to store only the name of the images in our  DB. Then we can find images successfully from that location. So we can simply use **models.charfield** to store images in the database, So now you can replace the models.URLField with models.CharField here.

File: users/models.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    country = models.ForeignKey(Country, on_delete=models.SET_NULL, blank=True, null=True)
```
Comment (TWijesooriya) [inline users/models.py]: 1. Make this country field required. 
2. Set on_delete = models.PROTECT

File: users/models.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    job_title = models.CharField(max_length=100)
+    experience = models.ForeignKey(Experience, on_delete=models.SET_NULL, blank=True, null=True)  # Changed "Experience" to "experience" to follow Python conventions
```
Comment (TWijesooriya) [inline users/models.py]: 1. Make this experience field required.
2. Set on_delete = models.PROTECT

File: users/models.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    website = models.URLField(max_length=200)
```
Comment (TWijesooriya) [inline users/models.py]: Make this website field optional.

File: users/models.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    description = models.TextField(max_length=1000)
+    total_score = models.IntegerField(default=0)
```
Comment (TWijesooriya) [inline users/models.py]: This field should be named as **profile_score**. Not **total_score**. Please use only the field names defined in the ER diagram

File: users/models.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+        
+class Country(models.Model):
+    label = models.CharField(max_length=100)
+    def __str__(self):
+        return self.label
+
+class Experience(models.Model):
+    label = models.TextField(max_length=500)
```
Comment (TWijesooriya) [inline users/models.py]: 1. We use models.TextField to store a text that has a large number of characters. And also, we don't need to specify a max_length property we using the models.TextField. In this scenario, the values for the label field are quite short. So you can store those values using a models.CharField.
2. And also make this field unique.

File: users/models.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+
+@receiver(post_save, sender=User)
+def save_user_profile(sender, instance, **kwargs):
+    instance.userprofile.save()
+
+
+
+class Interest(models.Model):
+    label = models.TextField(max_length=400,blank=True,null=True)
```
Comment (TWijesooriya) [inline users/models.py]: Make this field required and unique

File: users/models.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+class UserInterest(models.Model):
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    Interest = models.ForeignKey(Interest,on_delete=models.CASCADE, blank=True, null=True)
```
Comment (TWijesooriya) [inline users/models.py]: Make this field required. 

And also, here one user can have multiple interests. But at the same time, one use can not have the same interest more than once. So the user and interest together should be unique.

File: users/models.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
+from django.db import models
+from django.contrib.auth.models import User
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+        
+class Country(models.Model):
+    label = models.CharField(max_length=100)
```
Comment (TWijesooriya) [inline users/models.py]: Make this field unique

File: articles/models.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
+from django.db import models
+from django.contrib.auth.models import User
+from users.models import Interest
+
+
+class ArticleStatus(models.Model):
+    label = models.CharField(max_length=100)
```
Comment (TWijesooriya) [inline articles/models.py]: Make this field unique

File: articles/models.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    def __str__(self):
+        return self.label
+
+class Article(models.Model):
+    user = models.ForeignKey(User, on_delete=models.CASCADE,related_name='articles_created')
```
Comment (TWijesooriya) [inline articles/models.py]: Change the related_name as 'articles'

File: articles/models.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    status = models.ForeignKey(ArticleStatus,on_delete=models.SET_NULL,blank=True,null=True)
+    created_date = models.DateTimeField(auto_now_add=True)   
+    published_date = models.DateTimeField(blank=True, null=True)
+    title = models.CharField(max_length=100)
+    thumbnail_url = models.URLField(max_length=200)
+    content = models.TextField(max_length=20000,blank=True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. You don't need to specify the max_length property when using models.TextField.
2. Make this field required.

File: articles/models.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class ArticleStatus(models.Model):
+    label = models.CharField(max_length=100)
+    def __str__(self):
+        return self.label
+
+class Article(models.Model):
+    user = models.ForeignKey(User, on_delete=models.CASCADE,related_name='articles_created')
+    status = models.ForeignKey(ArticleStatus,on_delete=models.SET_NULL,blank=True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. Make this field required
2. set related_name to 'articles'
3. We don't need to set this field to null if the related status value is deleted from the **ArticleStatus** table. What we should do instead is, we should prevent delete values from **ArticleStatus** table, if it's values are being used in **Article** table. For that, we should set the value of **on_delete** as **"models.PROTECT"**.
4. Set the value of the **related_name** to "articles"

File: articles/models.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    created_date = models.DateTimeField(auto_now_add=True)   
+    published_date = models.DateTimeField(blank=True, null=True)
+    title = models.CharField(max_length=100)
+    thumbnail_url = models.URLField(max_length=200)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. I've mistakenly named this field as 'thumbnail_url' in our ER diagram. This field should be re-named as 'thumbnail'.
2. Make this field optional

File: articles/models.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    content = models.TextField(max_length=20000,blank=True,null=True)
+    category = models.ForeignKey(Interest,on_delete=models.SET_NULL,blank=True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. Make this field required.
2. Set the value of on_delete to "models.PROTECT"
3. set the value of  **related_name** to "articles"

File: articles/models.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    summary = models.TextField(max_length=500,blank=True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. Use models.Charfield here
2. Reduce the max_length to 250 characters

File: articles/models.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    current_reviewer = models.ForeignKey(User,on_delete=models.SET_NULL,blank=True,null=True,related_name='articles_reviewing')
+
+    def __str__(self):
+        return self.title
+
+
+class ArticleComment(models.Model):
+    user = models.ForeignKey(User,on_delete=models.CASCADE)
```
Comment (TWijesooriya) [inline articles/models.py]: Set the related_name to "comments"

File: articles/models.py [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    content = models.TextField(max_length = 500, blank = True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. Use models.CharField here
2. Make this field required

File: articles/models.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    article = models.ForeignKey(Article,on_delete=models.CASCADE)
```
Comment (TWijesooriya) [inline articles/models.py:30]: Set the related_name to "comments"

File: articles/models.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    def __str__(self):
+        return self.content
+
+
+class ReviewRequestStatus(models.Model):
+    label = models.CharField(max_length=100)
```
Comment (TWijesooriya) [inline articles/models.py]: Make this field unique

File: articles/models.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    def __str__(self):
+        return self.label
+
+class ReviewRequest(models.Model):
+    reviewer =  models.ForeignKey(User,on_delete=models.SET_NULL,blank=True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. Set on_delete to models.CASCADE
2. Make this field required
3. Set related_name to "review_requests"

File: articles/models.py [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+    article = models.ForeignKey(Article,on_delete=models.CASCADE)
```
Comment (TWijesooriya) [inline articles/models.py:42]: Set related_name to "review_requests"

File: articles/models.py [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    status = models.ForeignKey(ReviewRequestStatus,on_delete=models.SET_NULL,blank=True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. Set on_delete to models.PROTECT
2. Make this field required

File: articles/models.py [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+    declined_reason = models.TextField(max_length=500,blank=True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: Use models.CharField here

File: articles/models.py [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+    def __str__(self):
+        return f"{self.article.title} : status {self.status}"
+
+
+class Suggestion(models.Model):
+    review_request= models.ForeignKey(ReviewRequest,on_delete=models.CASCADE)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. This should be a one-to-one relationship.
2. Make this field unique

File: articles/models.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+    title = models.CharField(max_length=100)
+    content = models.TextField(max_length=1000,blank=True,null=True)
```
Comment (TWijesooriya) [inline articles/models.py]: 1. remove max_length property
2. Make this field required

**Review by TWijesooriya — CHANGES_REQUESTED** [review]
You have done such a good job. I appreciate it.  :heart: 

I've reviewed your code in detail and found some improvements to make.

Apart from those improvements consider these issues too

1. Use commit convensions
2. Miss spelled reqirements.txt
3. gitignore ".vsode" folder
4. Show all the models in admin panel using list display
5. Change the app name from users to userprofile

File: articles/models.py (hunk 14 above)
Comment (sanjueranga) [reply articles/models.py]: Is it still a url field?

**Comment by sanjueranga** [comment]
- Two fields can't have the same related name "article". I remember that you requested this last time also. I dont know how to do that.
-  Adding Fields to admin panel : I did this in "populating data branch"
- Everything else is done

File: articles/models.py (hunk 14 above)
Comment (TWijesooriya) [reply articles/models.py]: Glad you reminded it. Should be a charField

**Review by TWijesooriya — APPROVED** [review]

**Pull request merged**